src/storage/UpdateEdgeProcessor.cpp

```diff
@@ -0,0 +1,373 @@
+        // update the props and the record
+        int32_t retIndex = -1;
+        if (returnColumnsNum_ > 0) {
+            for (auto& prop : this->edgeContext_.props_) {
+                if (prop.prop_.name == field && prop.returned_) {
+                    retIndex = prop.retIndex_;
+                    break;
+                }
+            }
+        }
+        switch (expValue.which()) {
+            case VAR_INT64: {
+                auto v = boost::get<int64_t>(expValue);
+                updater_->setInt(field, v);
+                if (retIndex != -1) record_[retIndex] = v;
+                break;
+            }
+            case VAR_DOUBLE: {
+                auto v = boost::get<double>(expValue);
+                updater_->setDouble(field, v);
+                if (retIndex != -1) record_[retIndex] = v;
+                break;
+            }
+            case VAR_BOOL: {
+                auto v = boost::get<bool>(expValue);
+                updater_->setBool(field, v);
+                if (retIndex != -1) record_[retIndex] = v;
+                break;
+            }
+            case VAR_STR: {
+                auto v = boost::get<std::string>(expValue);
+                updater_->setString(field, v);
+                if (retIndex != -1) record_[retIndex] = v;
+                break;
+             }
+            default:
+                LOG(FATAL) << "Unknown VariantType: " << expValue.which();
+        }
+    }
+    return kvstore::ResultCode::SUCCEEDED;
+}
+
+
+std::vector<kvstore::KV> UpdateEdgeProcessor::processData() {
+   std::vector<kvstore::KV> data;
+   data.emplace_back(key_, std::move(updater_->encode()));
+
+   if (!data.empty() && returnColumnsNum_ > 0) {
+       RowWriter writer(nullptr);
+       for (auto v : record_) {
+           switch (v.which()) {
+               case VAR_INT64:
+                   writer << boost::get<int64_t>(v);
+                   break;
+               case VAR_DOUBLE:
+                   writer << boost::get<double>(v);
+                   break;
+               case VAR_BOOL:
+                   writer << boost::get<bool>(v);
+                   break;
+               case VAR_STR:
+                   writer << boost::get<std::string>(v);
+                   break;
+               default:
+                   LOG(FATAL) << "Unknown VariantType: " << v.which();
+           }
+       }
+       respData_ = writer.encode();
+   }
+   return data;
+}
+
+
+cpp2::ErrorCode UpdateEdgeProcessor::checkAndBuildContexts(
+        const cpp2::UpdateEdgeRequest& req) {
+    auto retCode = QueryBaseProcessor<cpp2::UpdateEdgeRequest,
+         cpp2::UpdateResponse>::checkAndBuildContexts(req);
+    if (retCode != cpp2::ErrorCode::SUCCEEDED) {
+        return retCode;
+    }
+
+    // build context of the update items
+    if (this->expCtx_ == nullptr) {
+        this->expCtx_ = std::make_unique<ExpressionContext>();
+    }
+    for (auto& item : req.get_update_items()) {
```

Comment:
What's purpose about this code?   I can't understand it

Reply:
There maybe some expression in the value of update items. Here is the context of getting the relevant expression.

Reply:
Great